core/src/main/java/org/elasticsearch/action/support/ActiveShardsObserver.java

```diff
@@ -0,0 +1,136 @@
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.action.support;
+
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.admin.indices.create.CreateIndexClusterStateUpdateRequest;
+import org.elasticsearch.cluster.ClusterChangedEvent;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.ClusterStateObserver;
+import org.elasticsearch.cluster.ack.ClusterStateUpdateResponse;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.component.AbstractComponent;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.unit.TimeValue;
+import org.elasticsearch.node.NodeClosedException;
+import org.elasticsearch.threadpool.ThreadPool;
+
+import java.util.function.BiFunction;
+
+/**
+ * This class provides primitives for waiting for a configured number of shards
+ * to become active before sending a response on an {@link ActionListener}.
+ */
+public class ActiveShardsObserver extends AbstractComponent {
+
+    private final ClusterService clusterService;
+    private final ThreadPool threadPool;
+
+    public ActiveShardsObserver(final Settings settings, final ClusterService clusterService, final ThreadPool threadPool) {
+        super(settings);
+        this.clusterService = clusterService;
+        this.threadPool = threadPool;
+    }
+
+    /**
+     * Creates an {@link ActionListener} that responds to a cluster state update upon index creation and
+     * waits on the specified number of active shards to be started before sending the action's response
+     * over its listener.
+     *
+     * @param request the index creation request
+     * @param actionListener the main listener that is listening for responses from the index creation event
+     * @param onResult a function that takes the cluster state update acknowledged flag and timed out flag as parameters
+     *                 and returns a response to be sent on the listener
+     * @return ActionListener for responding to index creation cluster state events and sending the action response
+     *         over the main listener, after waiting for the requested number of shards to be active
+     */
+    public <T> ActionListener<ClusterStateUpdateResponse> waitForActiveShards(final CreateIndexClusterStateUpdateRequest request,
```

Comment:
does it make sense for this method to generate a listener? if we really need this, let's reflect it in the name. Reading how it's used, I think it will be cleaner to have the relevant listener created by the caller and have this method do something:

```
new ActionListener<ClusterStateUpdateResponse>() {
       public void onResponse(ClusterStateUpdateResponse response) {
                      if (response.isAcknowledged()) {
                           waitForActiveShards(request, actionListener, onResult); <--- call it here.
                     } else {
                           actionListener.onResponse(response);
                     } 
      } 

     onFailure(..) { 
          ....
     } 

```

Also - double checking - do we need both actionListener and onResult ? it seems like a double callback structure. Maybe it's just a naming issue, but it seems we can keep this simple and just let this method tell use if it timed out or not? (given my suggestion to to create the wrapper listener on the outside?